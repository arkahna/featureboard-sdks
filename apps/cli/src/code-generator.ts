import { Tree, formatFiles, generateFiles } from '@nx/devkit'
import { printChanges } from 'nx/src/command-line/generate/generate'
import { FsTree, flushChanges } from 'nx/src/generators/tree'
import * as path from 'path'
import prompts from 'prompts'
import { FeatureDto } from './feature-dto'
import {
    getDotNetNameSpace,
    getDotNetNameSpaceFromNx,
    toDotNetType,
    toPascalCase,
} from './templates/functions'

export type TemplateType = 'dotnet-api'

type CodeGeneratorOptions = {
    templateType: TemplateType
    organizationName: string
    featureBoardKey?: string
    featureBoardBearerToken?: string
    featureBoardprojectName?: string

    dryRun: boolean
    quiet: boolean
    verbose: boolean
    interactive: boolean
}

export type CodeGeneratorOptionsPath = CodeGeneratorOptions & {
    outputPath: string
}
export type CodeGeneratorOptionsTree = CodeGeneratorOptions & {
    tree: Tree
    realitiveFilePath: string
}

export async function codeGenerator(
    options: CodeGeneratorOptionsTree,
): Promise<void>
export async function codeGenerator(
    options: CodeGeneratorOptionsTree,
): Promise<void>
export async function codeGenerator(
    options: CodeGeneratorOptionsPath | CodeGeneratorOptionsTree,
): Promise<void> {
    const features = await getFeatures(options)

    let tree: Tree | undefined
    let srcFolder = './'
    let runSaveChanges = true
    if ('tree' in options) {
        tree = options.tree
        srcFolder = options.realitiveFilePath
        runSaveChanges = false
    } else {
        tree = new FsTree(options.outputPath, options.verbose)
        runSaveChanges = true
    }

    const templateSpecificOptions: any = {}
    switch (options.templateType) {
        case 'dotnet-api':
            templateSpecificOptions.namespace =
                'tree' in options
                    ? await getDotNetNameSpaceFromNx(
                          tree,
                          options.realitiveFilePath,
                      )
                    : getDotNetNameSpace(options.outputPath)
    }

    generateFiles(
        tree,
        path.join(__dirname, 'templates', options.templateType),
        srcFolder,
        {
            toPascalCase,
            toDotNetType,
            ...templateSpecificOptions,
            features: features,
        },
    )

    await formatFiles(tree)
    if (runSaveChanges) saveChanges(tree, options)
}

async function getFeatures({
    organizationName,
    featureBoardprojectName,
    featureBoardKey,
    featureBoardBearerToken,
    interactive,
}: {
    organizationName: string
    interactive: boolean
    featureBoardKey?: string
    featureBoardBearerToken?: string
    featureBoardprojectName?: string
}): Promise<FeatureDto[]> {
    const headers: HeadersInit = {
        'X-Organization': organizationName,
    }

    if (featureBoardKey) headers['x-api-key'] = featureBoardKey
    else if (featureBoardBearerToken)
        headers.Authorization = `Bearer ${featureBoardBearerToken}`
    else throw new Error('Auth token not set')

    const result = await fetch(
        'https://api.featureboard.dev/projects?deep=true',
        {
            headers: headers,
        },
    ).then(async (response) => {
        if (response.ok) {
            return await response.json()
        }
        throw new Error(
            `Unable to reretrieve roject data from feature board. Status Code: ${
                response.status
            } (${response.statusText}), Content: ${await response.text()}`,
        )
    })

    let project = result.projects.find(
        (x: any) =>
            x.name.toLowerCase() ===
            featureBoardprojectName?.toLocaleLowerCase(),
    )
    if (!project && interactive) {
        if (result.projects.length === 1) {
            project = result.projects[0]
            console.log(
                `One project found. Setting feature board project to ${project.name}`,
            )
        } else {
            const promptResult = await prompts({
                type: 'select',
                name: 'project',
                message: `Pick feature board project?`,
                validate: (x) =>
                    result.projects.map((x: any) => x.name).includes(x),
                choices: result.projects.map((x: any) => ({
                    title: x.name,
                    value: x,
                })),
            })

            project = promptResult.project
        }
    }
    if (!project) throw new Error('Unable to locate Project')

    return project.features
}

function saveChanges(
    tree: Tree,
    options: {
        dryRun: boolean
        quiet: boolean
    },
) {
    const changes = tree.listChanges()
    if (!options.quiet) {
        printChanges(changes)
        if (changes.length === 0) {
            console.log(`\nFiles are the same no changes were made.`)
            return
        }
    }

    if (!options.dryRun && changes.length !== 0) {
        flushChanges(tree.root, changes)
        return
    }
    console.log(`\nNOTE: The "dryRun" flag means no changes were made.`)
}
