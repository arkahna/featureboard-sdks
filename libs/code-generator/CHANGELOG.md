# @featureboard/code-generator

## 0.3.6

### Patch Changes

- 59c8364: Added namespace validator to cli

## 0.3.5

### Patch Changes

- 69510ca: Fix code gen generator
  Add add target executor

## 0.3.4

### Patch Changes

- 3041a28: Move token management code to new project api-authentication and allow nx-plugin to make use of it.

## 0.3.3

### Patch Changes

- 9f1d891: Better error logging for unauthorized requests

## 0.3.2

### Patch Changes

- c0a2eab: Fix build process for NX plugin

## 0.3.1

### Patch Changes

- cdc1490: Change the display name of project to product

## 0.3.0

### Minor Changes

- 9e22fb1: Add typescript code generator template

## 0.2.1

### Patch Changes

- 5755d8f: Updated documentation and instructions

## 0.2.0

### Minor Changes

- bc4d42a: Add login, select-organization and account commands to CLI

## 0.1.1

### Patch Changes

- 41f499f: Fix package process

## 0.1.0

### Minor Changes

- 7587dae: Add support for modern bundlers

## 0.0.2

### Patch Changes

- [#42](https://github.com/arkahna/featureboard-sdks/pull/42) [`777b2f1`](https://github.com/arkahna/featureboard-sdks/commit/777b2f1556f718d3e5799878f83a8519e855cfa1) Thanks [@emlyn-arkahna](https://github.com/emlyn-arkahna)! - Update user messages
