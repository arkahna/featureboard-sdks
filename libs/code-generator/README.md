# FeatureBoard Code Generator

NX tool that generates FeatureBoard SDKs for different languages.

![npm](https://img.shields.io/npm/v/%40featureboard%2Fcode-generator?logo=npm) ![npm](https://img.shields.io/npm/dt/%40featureboard%2Fcode-generator) ![GitHub Workflow](https://img.shields.io/github/actions/workflow/status/arkahna/featureboard-sdks/main.yml?logo=github)

## Installation

```bash
npm add @featureboard/code-generator
```
```bash
yarn add @featureboard/code-generator
```
```bash
pnpm add @featureboard/code-generator
```

## Release Notes

Our changelog can be found on [GitHub](https://github.com/arkahna/featureboard-sdks/blob/main/libs/code-generator/CHANGELOG.md).
