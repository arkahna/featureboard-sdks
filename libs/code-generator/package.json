{
    "name": "@featureboard/code-generator",
    "version": "0.3.0",
    "main": "src/index.js",
    "license": "MIT",
    "sideEffects": false,
    "repository": {
        "type": "git",
        "url": "git+https://github.com/arkahna/featureboard-sdks.git"
    },
    "keywords": [
        "Feature management",
        "Feature toggles",
        "Feature flags"
    ],
    "bugs": {
        "url": "https://github.com/arkahna/featureboard-sdks/issues"
    },
    "homepage": "https://github.com/arkahna/featureboard-sdks/tree/master/libs/code-generator#readme"
}
