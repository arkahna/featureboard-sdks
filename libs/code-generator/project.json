{
    "name": "code-generator",
    "$schema": "../../node_modules/nx/schemas/project-schema.json",
    "sourceRoot": "libs/code-generator/src",
    "projectType": "library",
    "targets": {
        "build": {
            "executor": "@nx/js:tsc",
            "outputs": ["{options.outputPath}"],
            "options": {
                "outputPath": "dist/libs/code-generator",
                "tsConfig": "libs/code-generator/tsconfig.lib.json",
                "packageJson": "libs/code-generator/package.json",
                "main": "libs/code-generator/src/index.ts",
                "assets": [
                    "libs/code-generator/*.md",
                    {
                        "input": "./libs/code-generator/src",
                        "glob": "**/*.template",
                        "output": "./src"
                    }
                ]
            }
        },
        "lint": {
            "executor": "@nx/eslint:lint",
            "outputs": ["{options.outputFile}"],
            "options": {
                "lintFilePatterns": [
                    "libs/code-generator/**/*.ts",
                    "libs/code-generator/package.json"
                ]
            }
        }
    },
    "tags": []
}
