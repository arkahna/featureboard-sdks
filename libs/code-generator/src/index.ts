export { getUserOrganizations } from './lib/api/get-organizations'
export { getProjectFeatures } from './lib/api/get-project-features'
export type { FeatureBoardFeature } from './lib/api/get-project-features'
export { getProjects } from './lib/api/get-projects'
export { codeGenerator } from './lib/code-generator'
export type { CodeGeneratorOptions, Template } from './lib/code-generator'
export { queryFeatureBoard } from './lib/queryFeatureBoard'
export type { FeatureBoardAuth } from './lib/queryFeatureBoard'
export { FsTree, flushChanges, printChanges } from './lib/tree/tree'
export type { Tree } from './lib/tree/tree'
