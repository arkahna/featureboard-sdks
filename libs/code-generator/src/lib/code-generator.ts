import path from 'node:path'
import prompts from 'prompts'
import type {
    FeatureBoardFeature,
    FeatureBoardProject,
} from './api/get-project-features'
import { getProjectFeatures } from './api/get-project-features'
import { getProjects } from './api/get-projects'
import {
    getDotNetNameSpace,
    toDotNetType,
    toPascalCase,
} from './generators/dotnet-api/functions'
import { generateFiles } from './generators/dotnet-api/generate-files'
import type { FeatureBoardAuth } from './queryFeatureBoard'
import type { Tree } from './tree/tree'

export type Template = 'dotnet-api' | 'typescript'

export interface CodeGeneratorOptions {
    template: Template
    auth: FeatureBoardAuth
    apiEndpoint?: string
    featureBoardProjectName?: string

    interactive: boolean

    tree: Tree
    relativeFilePath: string
}

export async function codeGenerator(
    options: CodeGeneratorOptions,
): Promise<void> {
    const [project, features] = await getFeatures(options)
    if (!project || !features) {
        return
    }
    console.log(project)
    const relativeFilePath = options.relativeFilePath

    const templateSpecificOptions: any = {}
    switch (options.template) {
        case 'dotnet-api':
            templateSpecificOptions.namespace = getDotNetNameSpace(
                options.tree,
                relativeFilePath,
            )
    }

    await generateFiles(
        options.tree,
        path.join(__dirname, 'templates', options.template),
        options.relativeFilePath,
        {
            toPascalCase,
            toDotNetType,
            ...templateSpecificOptions,
            project,
            features: features,
        },
    )
}

async function getFeatures({
    featureBoardProjectName,
    auth,
    interactive,
    apiEndpoint = 'https://api.featureboard.app',
}: {
    interactive: boolean
    auth: FeatureBoardAuth
    featureBoardProjectName?: string
    apiEndpoint?: string
}): Promise<[null | FeatureBoardProject, null | FeatureBoardFeature[]]> {
    const projectResults = await getProjects(apiEndpoint, auth)

    let project = projectResults.projects.find(
        (x: any) =>
            x.name.toLowerCase() ===
            featureBoardProjectName?.toLocaleLowerCase(),
    )

    if (!project && interactive) {
        if (projectResults.projects.length === 1) {
            project = projectResults.projects[0]
            console.log(
                `One project found. Setting FeatureBoard project to ${project.name}`,
            )
        } else {
            const promptResult = await prompts({
                type: 'select',
                name: 'project',
                message: `Pick your FeatureBoard project?`,
                validate: (x) =>
                    projectResults.projects.map((x) => x.name).includes(x),
                choices: projectResults.projects.map((x) => ({
                    title: x.name,
                    value: x,
                })),
            })

            if (!promptResult) {
                return [null, null]
            }

            project = promptResult.project
        }
    }
    if (!project) throw new Error('Unable to locate Project')

    const featuresResult = await getProjectFeatures(apiEndpoint, project, auth)

    return [project, featuresResult.features]
}
