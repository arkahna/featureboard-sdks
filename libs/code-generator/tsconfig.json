{
    "extends": "../../tsconfig.settings.json",
    "compilerOptions": {
        "outDir": "./src",
        "rootDir": "./src",
        "types": [
            "node",
            "vitest/globals",
            "vitest/importMeta",
            "vite/client",
            "vitest"
        ]
    },
    "include": ["src/**/*.ts"],
    "exclude": ["node_modules", "tsc-out", "dist", "**/src/*.d.ts", "**/*.js"]
}
