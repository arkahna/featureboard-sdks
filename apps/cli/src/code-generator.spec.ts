import {
    Tree,
    addProjectConfiguration,
    joinPathFragments,
    readProjectConfiguration,
} from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import * as fs from 'fs/promises'
import fetchMock from 'jest-fetch-mock'
import * as path from 'path'

import { CodeGeneratorOptionsTree, codeGenerator } from './code-generator'

fetchMock.enableMocks()
describe('code-generator', () => {
    let tree: Tree
    let options: Omit<
        CodeGeneratorOptionsTree,
        'tree' | 'realitiveFilePath'
    > & { subFolder: string }
    describe('templateType: dotnet-api', () => {
        beforeEach(async () => {
            tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' })

            const root = 'apps/my-app'
            addProjectConfiguration(tree, 'my-app', {
                root: 'apps/my-app',
                projectType: 'application',
                targets: { 'my-target': { executor: 'nx:noop' } },
            })
            tree.write(
                joinPathFragments(root, 'Project.FeatureBoard.MyApp.csproj'),
                '',
            )

            options = {
                organizationName: 'featureboard',
                subFolder: './',
                templateType: 'dotnet-api',
                dryRun: false,
                quiet: false,
                verbose: true,
                interactive: false,
                featureBoardprojectName: 'FeatureBoard Service',
                featureBoardKey: 'This is totaly a key',
            }

            fetchMock.mockIf(
                /^https?:\/\/api.featureboard.dev.*$/,
                async (req) => {
                    const file = await fs.readFile(
                        path.join(__dirname, '../test-data/feature-board.json'),
                        {
                            encoding: 'utf8',
                        },
                    )
                    return {
                        status: 200,

                        body: file,
                    }
                },
            )
        })

        it('should run successfully', async () => {
            let project = readProjectConfiguration(tree, 'my-app')
            await codeGenerator({
                tree: tree,
                realitiveFilePath: joinPathFragments(
                    project.root,
                    options.subFolder,
                ),
                ...options,
            })

            project = readProjectConfiguration(tree, 'my-app')
            expect(project).toBeDefined()
        })

        it('should produce the expected features files', async () => {
            let project = readProjectConfiguration(tree, 'my-app')
            await codeGenerator({
                tree: tree,
                realitiveFilePath: joinPathFragments(
                    project.root,
                    options.subFolder,
                ),
                ...options,
            })

            project = readProjectConfiguration(tree, 'my-app')
            const files = tree
                .listChanges()
                .filter((x) => x.path.match(/^apps\/my-app\/.*$/))

            expect(files).toHaveLength(3)
            files.forEach((x) => {
                expect(x.content?.toString('utf-8')).toMatchSnapshot(x.path)
            })
        })
    })
})
