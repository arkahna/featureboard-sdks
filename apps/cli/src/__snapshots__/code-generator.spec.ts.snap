// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code-generator templateType: dotnet-api should produce the expected features files: apps/my-app/Features.cs 1`] = `
"using FeatureBoard.DotnetSdk.Attributes;
using FeatureBoard.DotnetSdk.Models;
using System.Runtime.Serialization;

namespace Project.FeatureBoard.MyApp;

// Features
public class Features : IFeatures
{
  [FeatureKeyName("organization-management")]
  public bool OrganizationManagement { get; set; }
  [FeatureKeyName("projects-write")]
  public bool ProjectsWrite { get; set; }
  [FeatureKeyName("environments-write")]
  public bool EnvironmentsWrite { get; set; }
  [FeatureKeyName("features-write")]
  public bool FeaturesWrite { get; set; }
  [FeatureKeyName("audiences-write")]
  public bool AudiencesWrite { get; set; }
  [FeatureKeyName("set-feature-availability")]
  public bool SetFeatureAvailability { get; set; }
  [FeatureKeyName("service-administration")]
  public bool ServiceAdministration { get; set; }
  [FeatureKeyName("webhooks-write")]
  public bool WebhooksWrite { get; set; }
  [FeatureKeyName("features-value-write")]
  public bool FeaturesValueWrite { get; set; }
  [FeatureKeyName("delete-organization")]
  public bool DeleteOrganization { get; set; }
  [FeatureKeyName("api-key-management")]
  public bool ApiKeyManagement { get; set; }
  [FeatureKeyName("webhooks")]
  public bool Webhooks { get; set; }
  [FeatureKeyName("cache-ttl")]
  public decimal CacheTtl { get; set; }
  [FeatureKeyName("live-updates-enabled")]
  public bool LiveUpdatesEnabled { get; set; }
  [FeatureKeyName("max-full-users")]
  public decimal MaxFullUsers { get; set; }
  [FeatureKeyName("max-limited-users")]
  public decimal MaxLimitedUsers { get; set; }
  [FeatureKeyName("max-projects")]
  public decimal MaxProjects { get; set; }
  [FeatureKeyName("webhook-type")]
  public WebhookType WebhookType { get; set; }
  [FeatureKeyName("organization-creation")]
  public bool OrganizationCreation { get; set; }
  [FeatureKeyName("environments-key-read")]
  public bool EnvironmentsKeyRead { get; set; }
}


// Options Enums

//Available webhook types 
public enum WebhookType
{
  [EnumMember(Value = "none")]
  None,
  [EnumMember(Value = "basic")]
  Basic,
  [EnumMember(Value = "advanced")]
  Advanced,
}

// Type Enums
public enum BooleanFeature
{
  OrganizationManagement,
  ProjectsWrite,
  EnvironmentsWrite,
  FeaturesWrite,
  AudiencesWrite,
  SetFeatureAvailability,
  ServiceAdministration,
  WebhooksWrite,
  FeaturesValueWrite,
  DeleteOrganization,
  ApiKeyManagement,
  Webhooks,
  LiveUpdatesEnabled,
  OrganizationCreation,
  EnvironmentsKeyRead,
}
public enum NumberFeature
{
  CacheTtl,
  MaxFullUsers,
  MaxLimitedUsers,
  MaxProjects,
}
public enum StringFeature
{
}

// Feature Gate attribute
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
public class FeatureFilterAttribute : FeatureFilterAttributeBase
{
  protected override string Feature { get; }
  protected override object DefaultValue { get; }
  protected override object? IsEqualTo { get; }
  private readonly Dictionary<string, string> _featureKeyLookup = new()
  {
    { "OrganizationManagement", "organization-management" },
    { "ProjectsWrite", "projects-write" },
    { "EnvironmentsWrite", "environments-write" },
    { "FeaturesWrite", "features-write" },
    { "AudiencesWrite", "audiences-write" },
    { "SetFeatureAvailability", "set-feature-availability" },
    { "ServiceAdministration", "service-administration" },
    { "WebhooksWrite", "webhooks-write" },
    { "FeaturesValueWrite", "features-value-write" },
    { "DeleteOrganization", "delete-organization" },
    { "ApiKeyManagement", "api-key-management" },
    { "Webhooks", "webhooks" },
    { "CacheTtl", "cache-ttl" },
    { "LiveUpdatesEnabled", "live-updates-enabled" },
    { "MaxFullUsers", "max-full-users" },
    { "MaxLimitedUsers", "max-limited-users" },
    { "MaxProjects", "max-projects" },
    { "OrganizationCreation", "organization-creation" },
    { "EnvironmentsKeyRead", "environments-key-read" },
  };

  public FeatureFilterAttribute(BooleanFeature feature, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(BooleanFeature feature, bool isEqualTo, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(StringFeature feature, string isEqualTo, string defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(NumberFeature feature, decimal isEqualTo, decimal defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(WebhookType isEqualTo, WebhookType defaultValue)
  {
    Feature = "webhook-type";
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

}
"
`;

exports[`code-generator templateType: dotnet-api should produce the expected features files: apps/my-app/Project.FeatureBoard.MyApp.csproj 1`] = `""`;

exports[`code-generator templateType: dotnet-api should produce the expected features files: apps/my-app/project.json 1`] = `
"{
  "name": "my-app",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "application",
  "targets": {
    "my-target": {
      "executor": "nx:noop"
    }
  }
}
"
`;
